rfcs/015 - TCP RFC.md

```diff
@@ -0,0 +1,196 @@
+
+## Expected SemBDD Impact
+
+This would be a minor update if Alan was post-1.0.0 as it adds new functionality without regressing on existing functionality.
+
+## Proposal
+
+A TCP connection has four essential states: open, closed-on-purpose, closed-on-error, and closed-by-timeout. While in the open state, data can be read from its output, written to its input, or explicitly closed as triggers by the user, while it can also unexpectedly transition to the other two closed states. The amount of data received at any particular point in time is completely arbitrary and up to both the originating machine and the local kernel, which makes it a good fit for the event system so the users' code is prompted when each chunk is ready to work with that it can continue with.
+
+The main issue is that Alan's event system is intentionally a static event system determined at compile time, so it is impossible to add new event handlers for each new socket that is opened by the users' code or has been received by the TCP server. The secondary issue is that the AVM's internal memory model would require an expensive conversion of that chunk from packed bytes to an internal `Array<int8>` type which, to simplify the memory model, is actually really an `Array<int64>` under the hood, which is very wasteful for a true byte array, so being able to avoid that conversion when possible would be good. Finally, the ergonomics of the API need to be both easy-to-understand and mesh well with the existing syntax while adding stream processing support at the same time.
+
+While changes to the language are on the table (and some of the alternatives will cover a few of them), the best overall solution (considering implementation complexity and cognitive complexity) doesn't need it. :)
+
+The clearest way to introduce the new syntax is to simply use it in an example similar to how it will be used by Anycloud:
+
+```ln
+from @std/tcpserver import connection, Connection
+from @std/tcp import connect, addContext, chunk, Chunk, Context, read, write, connClose, close, connError, connTimeout
+
+on connection fn (conn: Connection) {
+  const tunnel = connect('localhost', 8088);
+  conn.addContext(tunnel);
+  tunnel.addContext(conn);
+}
+
+on chunk fn (ctx: Context<Connection>) {
+  const c: Chunk = ctx.conn.read();
+  ctx.context.write(c);
+}
+
+on connClose fn (ctx: Context<Connection>) {
+  ctx.context.close();
+}
+
+on connError fn (ctx: Context<Connection>) {
+  ctx.context.close();
+}
+
+on connTimeout fn (ctx: Context<Connection>) {
+  ctx.context.close();
+}
+```
```

Comment:
What about listening on multiple ports, or closing down specific ports? It seems that this implementation still requires the remote to only connect to 1 port, at which point the Alan code will move the connection to a different port, which doesn't seem very flexible.

For example, say Alice wanted to implement a web server in Alan that allowed for both HTTP and SSH connections (just using them because I'm familiar with the packet specifications of both). To do this, she can only write the above, except for this change:
```ln
on connection fn (conn: Connection) {
  const c: Chunk = conn.read();
  const tunnel = c.getTunnel();
  if tunnel.isSome() {
    conn.addContext(tunnel);
    tunnel.addContext(conn);
  } else {
    conn.close();
  }
}
fn getTunnel(conn: Connection) {
  const chunk = conn.read();
  if chunk.startsWith('HTTP') {
    return some(connect('localhost', 8088));
  } else if chunk.startsWith('SSH') {
    return some(connect('localhost', 2222));
  } else {
    return none();
  }
}
```

Reply:
First, I attempted to address that here: https://github.com/alantech/alan/pull/467/files#diff-d20ad24b805e5d94f7929b3e2aed48bdad3f4613a80aea17d0b41b4346710665R73

> Whether or not that autoselection (of the listen port) sticks around long-term is still up in the air.

I was explicitly not addressing the situation you bring up as out-of-scope here. It *should* be possible to do something like telling the AVM what `tcp` ports (plural) to listen to and then the `conn` object has a `port` field that you can key on (or maybe a `serverNum` so the port itself is opaque but "first server" vs "second server" is clear in your code) that can then be tackled. And this doesn't seem like it would require ripping everything of this design out, just something that wouldn't be tackled at first.

For the exact situation where the same port handles both HTTP and SSH connections, since `addContext` *can* be called in the `chunk` events as described further down: https://github.com/alantech/alan/pull/467/files#diff-d20ad24b805e5d94f7929b3e2aed48bdad3f4613a80aea17d0b41b4346710665R89

That means that the user could create a type where one of the fields is `Maybe<Connection>` and the other is `Maybe<Array<Chunk>>` (or something) and spinning up the correct child connection for an HTTP server or an SSH server can be deferred until later, and the `chunk` handler function just checks whether the context's `Maybe<Connection>` is defined or not for whether or not it will write the chunk to an established connection, or attempt to probe which kind of connection it is. If it's streaming in, it can also check if `Maybe<Array<Chunk>>` is defined or not to first stream the already buffered array of chunks in before streaming the current chunk, or if it's just going to write the current chunk only.

This also makes it invulnerable to the possible but unlikely situation that the client is super slow and the first chunk you get is just "HT", for instance, and you can't identify it by the first chunk of data alone.

So it wouldn't need to have access to the `read` function during establishing the connection (nor would that be expected to already have a chunk). The current design could work for that scenario, too.

Reply:
Gotcha, makes sense. Another question I've thought of, but is rather out-of-scope for this RFC: how will `@std/http` know which port to listen to depending on usage in `anycloud` or not? We could define the default port to 8088, but what if users want to listen to port 80 instead?

Reply:
So an Alan application itself is not deployed through Anycloud. It deploys directly with `alan deploy` and this problem is moot. The only HTTP server is actually listening on port 443 as an HTTPS server with a self-signed cert that is connected to Cloudflare where the real cert lives.

Reply:
If you just run an Alan application with `alan run myapp.agz` it will listen on port 8000, intentionally wrong for Anycloud to discourage trying to needlessly wrap your Alan app in Docker and another Alan app.